src/Std/Data/DTreeMap/Lemmas.lean

```diff
@@ -2946,6 +2959,139 @@ theorem minKey?_alter_eq_self [TransCmp cmp] {k f} :
 
 end Const
 
+theorem minKey_eq_get_minKey? [TransCmp cmp] {he} :
+    t.minKey he = t.minKey?.get (isSome_minKey?_iff_isEmpty_eq_false.mpr he) :=
+  Impl.minKey_eq_get_minKey? t.wf
+
+theorem minKey?_eq_some_minKey [TransCmp cmp] {he} :
+    t.minKey? = some (t.minKey he) :=
+  Impl.minKey?_eq_some_minKey t.wf
+
+theorem minKey_eq_iff_getKey?_eq_self_and_forall [TransCmp cmp] {he km} :
+    t.minKey he = km ↔ t.getKey? km = some km ∧ ∀ k ∈ t, (cmp km k).isLE :=
+  Impl.minKey_eq_iff_getKey?_eq_self_and_forall t.wf
+
+theorem minKey_eq_some_iff_mem_and_forall [TransCmp cmp] [LawfulEqCmp cmp] {he km} :
+    t.minKey he = km ↔ km ∈ t ∧ ∀ k ∈ t, (cmp km k).isLE :=
+  Impl.minKey_eq_some_iff_mem_and_forall t.wf
+
+theorem minKey_insert [TransCmp cmp] {k v} :
+    (t.insert k v).minKey isEmpty_insert =
+      (t.minKey?).elim k fun k' => if cmp k k'|>.isLE then k else k' :=
+  Impl.minKey_insert t.wf
+
+theorem minKey_insert_le_minKey [TransCmp cmp] {k v he} :
+    cmp (t.insert k v |>.minKey isEmpty_insert) (t.minKey he) |>.isLE :=
+  Impl.minKey_insert_le_minKey t.wf
+
+theorem minKey_insert_le_self [TransCmp cmp] {k v} :
+    cmp (t.insert k v |>.minKey isEmpty_insert) k |>.isLE :=
+  Impl.minKey_insert_le_self t.wf
+
+theorem contains_minKey [TransCmp cmp] {he} :
+    t.contains (t.minKey he) :=
+  Impl.contains_minKey t.wf
+
+theorem minKey_mem [TransCmp cmp] {he} :
+    t.minKey he ∈ t :=
+  Impl.minKey_mem t.wf
+
+theorem minKey_le_of_contains [TransCmp cmp] {k} (hc : t.contains k) :
+    cmp (t.minKey <| isEmpty_eq_false_iff_exists_contains_eq_true.mpr ⟨k, hc⟩) k |>.isLE :=
+  Impl.minKey_le_of_contains t.wf hc
+
+theorem minKey_le_of_mem [TransCmp cmp] {k} (hc : k ∈ t) :
+    cmp (t.minKey <| isEmpty_eq_false_iff_exists_contains_eq_true.mpr ⟨k, hc⟩) k |>.isLE :=
+  Impl.minKey_le_of_mem t.wf hc
+
+theorem le_minKey [TransCmp cmp] {k he} :
+    (cmp k (t.minKey he)).isLE ↔ (∀ k', k' ∈ t → (cmp k k').isLE) :=
+  Impl.le_minKey t.wf
+
+theorem getKey?_minKey [TransCmp cmp] {he} :
```

Comment:
Many good `simp` lemmas around this part I think.

Reply:
So true -- I also added `@[simp]` to `minKey_erase_eq_iff_not_compare_eq_minKey` and `minKey_modify`.